xgboost_ray/data_sources/dask.py

```diff
@@ -0,0 +1,153 @@
+from collections import defaultdict
+from typing import Any, Optional, Sequence, Dict, Union, Tuple
+
+import pandas as pd
+
+import ray
+from ray.actor import ActorHandle
+
+from xgboost_ray.data_sources._distributed import \
+    assign_partitions_to_actors, get_actor_rank_ips
+from xgboost_ray.data_sources.data_source import DataSource, RayFileType
+
+try:
+    import dask  # noqa: F401
+    from ray.util.dask import ray_dask_get
+    DASK_INSTALLED = True
+except ImportError:
+    DASK_INSTALLED = False
+
+
+def _assert_dask_installed():
+    if not DASK_INSTALLED:
+        raise RuntimeError(
+            "Tried to use Dask as a data source, but dask is not "
+            "installed. This function shouldn't have been called. "
+            "\nFIX THIS by installing dask: `pip install dask`. "
+            "\nPlease also raise an issue on our GitHub: "
+            "https://github.com/ray-project/xgboost_ray as this part of "
+            "the code should not have been reached.")
+
+
+class Dask(DataSource):
+    """Read from distributed Dask dataframe.
+
+    A `Dask dataframe <https://docs.dask.org/en/latest/dataframe.html>`_
+    is a distributed drop-in replacement for pandas.
+
+    Dask dataframes are stored on multiple actors, making them
+    suitable for distributed loading.
+    """
+    supports_central_loading = True
+    supports_distributed_loading = True
+
+    @staticmethod
+    def is_data_type(data: Any,
+                     filetype: Optional[RayFileType] = None) -> bool:
+        if not DASK_INSTALLED:
+            return False
+        from dask.dataframe import DataFrame as DaskDataFrame, \
+            Series as DaskSeries
+
+        return isinstance(data, (DaskDataFrame, DaskSeries))
+
+    @staticmethod
+    def load_data(
+            data: Any,  # dask.pandas.DataFrame
+            ignore: Optional[Sequence[str]] = None,
+            indices: Optional[Union[Sequence[int], Sequence[int]]] = None,
+            **kwargs) -> pd.DataFrame:
+        _assert_dask_installed()
+
+        import dask.dataframe as dd
+
+        if indices is not None and len(indices) > 0 and isinstance(
+                indices[0], Tuple):
+            # We got a list of partition IDs belonging to Dask partitions
+            return dd.concat(
+                [data.partitions[i] for (i, ) in indices]).compute()
+
+        # Dask does not support iloc() for row selection, so we have to
+        # compute a local pandas dataframe first
+        local_df = data.compute()
```

Comment:
will this materialize the full distributed dataframe?

Reply:
Yes. However, this part of the code concerns the local loading - i.e. loading of either a partition or centralized loading, where access to the whole dataframe is assumed anyway. This is the part where we end up with a pandas dataframe.